lib/rubocop/cli/command/json_schema.rb

```diff
@@ -0,0 +1,187 @@
+# frozen_string_literal: true
+
+module RuboCop
+  class CLI
+    module Command
+      # Output a JSON Schema file for the for the current directory's RuboCop
+      # configuration file.
+      #
+      # @api private
+      class JSONSchema < Base
+        # Cop attributes that shouldn't appear in a user config
+        NON_USER_COP_KEYS = %w[
+          Description StyleGuide Reference Safe SafeAutoCorrect VersionAdded
+          VersionChanged
+        ].freeze
+
+        self.command_name = :json_schema
+
+        def initialize(env)
+          super
+
+          # we need "require" so can't just do "@config_store.force_default_config!"
+          stripped_config = build_stripped_config
+          @config = load_stripped_config(stripped_config)
+        end
+
+        def run
+          schema = build_json_schema
+          puts JSON.pretty_generate(schema)
+        end
+
+        def build_json_schema
+          cops = RuboCop::Cop::Registry.global
+          department_definitions = build_department_definitions(cops)
+          cop_definitions = cops.to_a.map do |cop|
+            [cop.cop_name, build_cop_definition(cop)]
+          end.to_h
+
+          template_file = File.join(File.dirname(__FILE__), 'json_schema_template.yml')
+          schema = YAML.safe_load(File.read(template_file))
+          schema['properties'].merge!(department_definitions)
+          schema['properties'].merge!(cop_definitions)
+          schema
+        end
+
+        private
+
+        # Take just the "require" part of the user config to get RuboCop to load
+        # custom cops without also loading the user's settings.
+        def build_stripped_config
+          path = ConfigLoader.configuration_file_for(Dir.pwd)
+          hash = ConfigLoader.load_yaml_configuration(path)
+          hash.slice('require')
+        end
+
+        # RuboCop config init is tightly coupled to file loading, so we'll write
+        # our stripped-down config into a tempfile and then read it.
+        def load_stripped_config(stripped_config)
+          require 'tempfile'
+
+          Tempfile.open('rubocop_config_temp') do |f|
+            f.write(stripped_config.to_yaml)
+            f.rewind
+            begin
+              # silence RuboCop's "pending" warning
+              old_verbose = $VERBOSE
+              $VERBOSE = nil
+              ConfigLoader.configuration_from_file(f.path)
+            ensure
+              $VERBOSE = old_verbose
+            end
+          end
+        end
```

Comment:
All of this is just to load the default config for RuboCop and any required extensions.

Any ideas on better better ways to accomplish this?

Reply:
@jonas054 might have some ideas. He's our master of configuration. :-)

Reply:
No, I can't think of any better solution. Your solution seems well researched, @dleavitt. Also, this is new code we're talking about, not changing anything in existing configuration loading logic, so I don't foresee any problems.

Reply:
@jonas054 Thanks! I've actually had some trouble getting it to play well with the frozen registry in the tests - specifically getting `ConfigLoader.load_file` to be deterministic.

The first time it is run in a test it seems to add all cops specified in the config `require` line, but calls in subsequent tests populate the registry _only_ with the default built-in Cops and not with additional `require`'d ones. It's specifically related to the order the tests are run in.

Doing `RuboCop::ConfigLoader.clear_options` and similar do not seem to help. Removing the `:restore_registry` meta and unfreezing the registry _does_ make the tests pass (though obvs isn't the right solution.)

Minimal repro:

```ruby
# frozen_string_literal: true

RSpec.describe RuboCop::CLI::Command::JSONSchema, :isolated_environment, :restore_registry do
  include FileHelper

  3.times do
    # contents of file are always the same but performance cops are only loaded on the first run
    it 'is deterministic' do
      create_file('.rubocop.yml', 'require: [rubocop-performance]')
      
      path = RuboCop::ConfigLoader.configuration_file_for(Dir.pwd)
      RuboCop::ConfigLoader.configuration_from_file(path)
      # RuboCop::ConfigStore.new.for_pwd, #load_file, etc give same result

      # 520 built in cops, 49 performance cops
      # always 569 on first run and 520 subsequently
      expect(RuboCop::Cop::Registry.global.length).to eq 569
    end
  end
end

```

Reply:
Yes, I know what you mean. I've tried out your minimal repro now, but was unable to find a solution. I've come across similar problems before, and something I concluded then was, that it seems impossible to make a spec example that requires something and then have that require "undone". It's just not possible in Ruby, I think.

So my suggestion is that you try to work around, or avoid, the problem, rather than solve it. Sorry I can't be more helpful.

Reply:
No problem, that's helpful too.